File: russian-doll-envelopes.java
//TC: O(n^2)
//SC: O(n)
//running on leetcode: yes
class Solution {
    public int maxEnvelopes(int[][] envelopes) {
        if(envelopes==null || envelopes.length==0) return 0;
        int n= envelopes.length;
        Arrays.sort(envelopes, (a,b) -> a[0] - b[0]);
        int[] dp=new int[n];
        int max=1;
        Arrays.fill(dp,1);
        for(int i=1; i<n; i++){
            for (int j=0; j<i; j++){
                if(envelopes[i][0]>envelopes[j][0] && envelopes[i][1] > envelopes[j][1]){
                    dp[i]=Math.max(dp[i],1 + dp[j]);
                    max=Math.max(max, dp[i]);
                }
            }
        }
        return max;
    }
}

File: russian-doll-envelopes.py
# TC: O(nlogn) | SC: O(n)
class Solution:
    def maxEnvelopes(self, envelopes: List[List[int]]) -> int:
        envelopes.sort(key = lambda x : (x[0], -x[1]))

        def lis(nums):
            n = len(nums)
            if n<1: return 0
            maxSub = [nums[0]]
            for num in nums[1:]:
                i = bisect_left(maxSub, num)
                if i == len(maxSub):    maxSub.append(num)
                else: maxSub[i] = num

            return len(maxSub)

        return lis([x[1] for x in envelopes])